rollup/internal/controller/watcher/batch_proposer.go

```diff
@@ -155,6 +155,28 @@ func (p *BatchProposer) TryProposeBatch() {
 }
 
 func (p *BatchProposer) updateDBBatchInfo(batch *encoding.Batch, codecVersion encoding.CodecVersion, metrics utils.BatchMetrics) error {
+	for {
+		compatible, err := utils.CheckBatchCompressedDataCompatibility(batch, codecVersion)
+		if err != nil {
+			log.Error("Failed to check batch compressed data compatibility", "codecVersion", codecVersion, "err", err)
+			return err
+		}
+
+		if compatible {
+			break
+		}
+
+		if len(batch.Chunks) == 1 {
+			log.Error("Cannot truncate batch with only 1 chunk for compatibility", "start block number", batch.Chunks[0].Blocks[0].Header.Number.Uint64(),
+				"end block number", batch.Chunks[0].Blocks[len(batch.Chunks[0].Blocks)-1].Header.Number.Uint64())
+			return errors.New("cannot truncate batch with only 1 chunk for compatibility")
+		}
+
+		log.Info("Batch not compatible with compressed data, removing last chunk")
+		p.compressedDataCompatibilityBreachTotal.Inc()
+		batch.Chunks = batch.Chunks[:len(batch.Chunks)-1]
+	}
+
 	err := p.db.Transaction(func(dbTX *gorm.DB) error {
 		dbBatch, dbErr := p.batchOrm.InsertBatch(p.ctx, batch, codecVersion, metrics, dbTX)
```

Comment:
`metrics` will be incorrect

Reply:
I considered this when implementing, and after revisiting, submitting only once makes more sense. will change it.

Reply:
No that's not my point. This `metrics` will be incorrect:

```
p.batchOrm.InsertBatch(p.ctx, batch, codecVersion, metrics, dbTX)
```

Reply:
nice catch. fixed in [9b7f048](https://github.com/scroll-tech/scroll/pull/1431/commits/9b7f048b93100f3712e6315a5828584729af4536).

the commit also simplified the related logic a bit.